nifi-nar-bundles/nifi-deltalake-bundle/nifi-deltalake-processors/src/main/java/org/apache/nifi/processors/deltalake/service/DeltaTableService.java

```diff
@@ -0,0 +1,129 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.processors.deltalake.service;
+
+import io.delta.standalone.Operation;
+import io.delta.standalone.OptimisticTransaction;
+import io.delta.standalone.actions.AddFile;
+import io.delta.standalone.actions.FileAction;
+import io.delta.standalone.actions.Metadata;
+import io.delta.standalone.actions.RemoveFile;
+import io.delta.standalone.data.CloseableIterator;
+import io.delta.standalone.types.StructType;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.nifi.processors.deltalake.storage.StorageAdapter;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+public class DeltaTableService {
+
+    private final static String DELTA_TABLE_NAME = "delta-table";
+    private final StructType structType;
+    private final StorageAdapter storageAdapter;
+
+    private OptimisticTransaction transaction;
+    private List<FileAction> fileUpdates;
+    private Metadata deltaTableMetadata;
+
+    public DeltaTableService(StorageAdapter storageAdapter, StructType structType) {
+        this.storageAdapter = storageAdapter;
+        this.structType = structType;
+    }
+
+    public void startTransaction() {
+        fileUpdates = new ArrayList<>();
+        deltaTableMetadata = createMetadata();
+        transaction = storageAdapter.getDeltaLog().startTransaction();
+    }
+
+    public void finishTransaction() {
+        if (!fileUpdates.isEmpty()) {
+            transaction.updateMetadata(deltaTableMetadata);
+            transaction.commit(fileUpdates, new Operation(Operation.Name.UPDATE), storageAdapter.getEngineInfo());
+            fileUpdates = Collections.emptyList();
+        }
+    }
+
+    public boolean deltaTableExists() {
+        return storageAdapter.getDeltaLog().tableExists();
+    }
+
+    public void addFilesToDeltaTable(List<AddFile> newFiles) {
+        if (!newFiles.isEmpty()) {
+            fileUpdates.addAll(newFiles);
+        }
+    }
+
+    public void removeFilesFromDeltaTable(List<RemoveFile> removeFileList) {
+        if (!removeFileList.isEmpty()) {
+            fileUpdates.addAll(removeFileList);
+        }
+    }
+
+    public void createNewDeltaTable(List<AddFile> newFiles) {
```

Comment:
The fileUpdates variable is null when a new table is created. The variable needs to be initialized as it is done 
 e.g. in startTransaction() method.
Or even startTransaction() can be invoked from this method as well.

Reply:
Thanks, updated.